src/irmin/store.ml

```diff
@@ -903,6 +903,54 @@ module Make (P: S.PRIVATE) = struct
         | _ -> Lwt.return_none
       ) g
 
+  module Heap = Bheap.Make(struct
+      type t = commit * int
+      let compare c1 c2 =
+         Int64.compare
+          (Info.date (Commit.info (fst c1)))
+          (Info.date (Commit.info (fst c2)))
+    end)
+
+  let last_modified ?depth ?(number = 1) t key =
+    Head.get t >>= fun commit ->
+    history t >>= fun history ->
+    let heap = Heap.create 5 in
+    let () = Heap.add heap (commit, 0) in
+    let rec search acc =
+      if Heap.is_empty heap || List.length acc = number then Lwt.return acc
+      else
+        let current, current_depth = Heap.pop_maximum heap in
+        let parents = History.pred history current in
+        of_commit current >>= fun store ->
+        find store key >>= fun current_value ->
+        if List.length parents = 0
+        then (
+          if current_value <> None
+          then Lwt.return (current :: acc)
+          else Lwt.return acc)
+        else
+          let max_depth = match depth with
+          | Some depth -> current_depth >= depth
+          | None -> false
+          in
+          Lwt_list.for_all_p
+            (fun commit ->
+              let () =
+                if not max_depth then
+                  Heap.add heap (commit, current_depth + 1)
+              in
+              of_commit commit >>= fun store ->
+              find store key >|= fun e ->
+              match (e, current_value) with
+              | Some x, Some y -> x != y
+              | Some _, None -> true
+              | _, _ -> false )
+            (History.pred history current)
```

Comment:
Same here, just use `Commit.parents current` and use `Commit.of_hash` in the callback. This will pose the question of what to do with shallow commits; I guess we can just ignore them.